docs(book): Add page on storage (#813)
future-proof-iot/RIOT-rs: book/src/storage.md

# Persistent Storage

Ariel OS supports persistent storage on a number of boards.

> It provides a key-value pair store based on [sequential-storage].

Author: I'm fine with replacing the dash for an en-dash.

[Sequential storage calls it a key–value pair store](https://docs.rs/sequential-storage/4.0.1/sequential_storage/index.html) and our [docs call it the same](https://ariel-os.github.io/ariel-os/dev/docs/api/ariel_os/storage/index.html). I want to keep it the same here for consistency.

The storage sets up the flash storage required.

## Use cases

Persistent storage can be used to store configuration and other variables

> across reboots and power interruptions on the microcontroller.

Author: I think this is sufficiently explained in the [Flash Requirements](https://github.com/ariel-os/ariel-os/pull/813/files#diff-4e4c279bc207db356a6dfb469c04b36f626e1bf3b7f2fa0f3a61c90d660d5ef7R32).

## Usage

### Enabling the storage module

The storage module can be enabled by [selecting the `sw/storage` laze module][laze-modules-book].

### Using the storage module

> The storage module provides the necessary getters and setters

Author: Added

> to interact with the storage backend

Author: Fixed

Keys are always `&str` typed.

> Values can have any type,

Author: You're right, clarified this in the text

but care must be taken to always use the same type with the same key.

See the [example][storage-example-repo] for details on the usage.

## Flash requirements

The storage module requires at least two flash pages.
These are allocated after the rodata in the flash when the module is enabled.

> 

Author: I agree, added some clarification.

> Updating the firmware can move and invalidate the storage pages
  when the firmware size differs from the previous version

> > Currently all flash pages must have an uniform size.

Author: Fixed

> 

Author: I wrote a bit, let me know if it is too much hand waving

> 

Author: That type of durability. [Sequential-storage claims to be power-fail safe](https://github.com/tweedegolf/sequential-storage/blob/master/README.md?plain=1#L65-L66) in the features. Is that what you have in mind here to be documented?

Something along the lines of:

> The storage module is protected via Sequential storage against corruption of individual key-value pairs. Consistency across writes to key-value pairs is not guaranteed.

> 

Author: From what I can find, the storage can always recover from any power failure. What I can not find in the documentation is if a key-value pair can end up in an invalid state. e.g. is it possible via an insert on an already existing item to get into a state where the newly written state is corrupted and the old state is already invalidated, when a power loss happens at the wrong moment. This would mean the whole storage is still recoverable, but that one item is now gone.

# Persistent Storage

Ariel OS supports persistent storage on a number of boards.
It provides a key–value pair store based on [sequential-storage].
The storage sets up the flash storage required.

## Use Cases

Persistent storage can be used to store configuration and other variables
across reboots and power interruptions on the microcontroller.

## Usage

### Enabling the Storage Module

The storage module can be enabled by [selecting the `sw/storage` laze module][laze-modules-book].

### Using the Storage Module

The [storage module] provides the necessary getters and setters
to interact with the storage backend.

Keys are always `&str` typed.
Values are required to implement the [`serde::Serialize`][serde-serialize]

> and [`serde::Deserialize`][serde-deserialize] traits.

Reviewer: ```suggestion
and [`serde::Deserialize`][serde-deserialize] traits.
Under the hood, currently the values are serialized using [postcard](https://github.com/jamesmunns/postcard).
```

Author: I think this is wat @ROMemories considers to be an implementation detail and doesn't belong here.

Reviewer: hmok. As a reader I'd be interested in that but don't insist.

> Care must be taken to always use the same value type with the same key.

Reviewer: ```suggestion
Care must be taken to always read a key with the same value type that it was written before.
While using a different value type for reading than for writing is never unsafe, it might result in bogus data.
```

which enables the `storage` Cargo feature.

### Using the Storage Module

The [storage module] provides the necessary getters and setters
to interact with the storage backend.

Keys are always `&str` typed.
Values are required to implement the [`serde::Serialize`][serde-serialize]
and [`serde::Deserialize`][serde-deserialize] traits.
Under the hood, currently the values are serialized using [postcard].

Care must be taken to always read a key with the same value type
that it was written before.
While using a different value type for reading than for writing is never unsafe,
it might result in bogus data.

See the [example][storage-example-repo] for details on the usage.

## Flash Requirements

The storage module requires at least two flash pages.
The effective storage space available is `(N - 1) * PAGE_SIZE`,
where `N` is the number of flash pages allocated.

> Currently storage is only supported on flash whose pages have a uniform size.

These pages are allocated by Ariel OS after the `rodata` section in the flash
when the module is enabled.

> Updating the firmware can move and invalidate the storage pages

>   when the firmware size differs from the previous version.

Reviewer 2: I was about to add "and across firmware updates" on top until I found this.

Is this something we want to fix before the release? This is a severe confusion source. (Also, it means that as long as that is in, we don't really have to bother with format stability, because chances are its' gone soon anyway, and only likely to not move if you're tweaking minor aspects).

NOR flash has limited endurance.
When writing applications using the storage module,

> care must be taken to limit the writes to a reasonable amount.

Reviewer 2: ```suggestion
care must be taken to limit the writes to a reasonable amount,
especially at the startup when there is the danger of a reboot due to a crash that causes endless writing.
```

See the [example][storage-example-repo] for details on the usage.

### Durability and Corruption

The underlying [sequential-storage] crate guarantees that the storage can be repaired
after power failure during operations.
Some corruption leads to unrecoverable data that cannot be repaired.
The repair will always make sure that the flash state is recoverd,

> so that any next operation should succeed.

Author: I've kept the 'should'.
